### PR TITLE
Added travis ci support for modern node versions

### PR DESCRIPTION
**Pull request by amilajack**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,20 +1,27 @@
...
+
 cache:
   directories:
     - node_modules
+
 node_js:
-  - 0.1
-  - 4.3
+  - node
   - stable
```
Comment (keithamus) [inline .travis.yml]: This is not quite right.

 - `node` is synonymous with `stable`
 - We want to keep testing LTS versions (while 4.3 is outdated, we should update this to 4.5 or maybe just 4)
 - We still want to test Node 0.1, until we stop and that will be a breaking change

Comment (keithamus) [inline .travis.yml]: I think the right setup is:

```yaml
node_js
 - 0.1 # remove this with the next breaking change
 - lts/*
 - stable
```

**Comment by amilajack** [comment]
Fixed

**Comment by keithamus** [comment]
LGTM. @chaijs/type-detect team, anyone want to pass comment?

**Comment by keithamus** [comment]
Ahh... for some reason `lts/*` didn't work. NVM docs say it should do: https://github.com/creationix/nvm#long-term-support. Not sure whats going on here 😞

**Comment by keithamus** [comment]
So nvm 0.31.4 added LTS support (https://github.com/creationix/nvm/releases/tag/v0.31.3), travis is using 0.31.0. There's a PR on travis to update it to 0.31.4 here: https://github.com/travis-ci/travis-build/pull/821.

**Comment by davelosert** [comment]
@keithamus : That PR was merged as it seems - does it fix the issue?

**Comment by keithamus** [comment]
@davelosert @amilajack we need https://github.com/travis-ci/travis-build/pull/821/files to be merged, then this will work.

**Comment by davelosert** [comment]
@keithamus Sorry - i saw the green lights on the bottom of that PR and thought it was merged - but it is still open. Nevermind then - we wait for it to be merged. :+1: 

Besides that good work @amilajack :)

**Comment by lucasfcosta** [comment]
@davelosert @keithamus Yay, that PR was merged!
Let's get this party going! 😄

**Comment by keithamus** [comment]
Its working now. Just need sign off from one of @davelosert @lucasfcosta @meeber

**Comment by meeber** [comment]
Do we want to align Node version numbers with base Chai? If so, then I'd recommend:

- 0.10 (which as noted will soon be phased out)
- 0.12
- 4
- 6

I'm not a fan of the `node` alias (replacement for `stable` which is now deprecated) for this type of project because the moment a new major version is released, our CI automatically stops testing the previous version even if it's still supported. I'm not a fan of `lts/*` because there are periods of time when there are two LTS versions being supported but that alias only brings in the latest one. It's my current feeling that we should explicitly list all the version numbers currently supported.

**Comment by keithamus** [comment]
@meeber okay, some valid points there - but I also think we need aliases. For example, the Argon LTS release is a moving target, it started as `4.2.0`and it's now `4.5.0`. We could, I suppose, explicitly add `0.10` and `0.12`. I think `stable` is worthwhile keeping, because it would be a nice assurance that new commits will run on the latest stable build. Perhaps we could do the following then @meeber:

```yml
node_js
 - 0.10 # to be removed 2016-10-01
 - 0.12 # to be removed 2016-12-31
 - lts/argon # to be removed 2018-04-01
 - lts/* # (never remove this)
 - stable # (never remove this)
```

**Comment by meeber** [comment]
The major version of a LTS release won't change, so `lts/argon` and `4` are equivalent. I believe the simpler numeric version is clearer, since (at least in my experience) people typically use the numeric version in normal discussion and use. I install "Node 4", I don't install "Node Argon". I ask a colleague if a language feature works in "Node 4", not in "Node Argon".

I don't have a problem including `lts/*` and `node` (`stable` is deprecated). However, I think the current numeric version should also be included, so that testing for a non-LTS version isn't automatically dropped the moment the next version is released. Typically there's a couple month period where the previous non-LTS version is still supported, as was the case with v5 after v6 was released.

By the way, I like having comments.

Counter-proposal:

```js
node_js
 - 0.10 # to be removed 2016-10-01
 - 0.12 # to be removed 2016-12-31
 - 4 # to be removed 2018-04-01
 - 6 # to be removed 2019-04-01
 - lts/* # (never remove this)
 - stable # (never remove this)
```

**Comment by keithamus** [comment]
Thoughts from others on this?

**Comment by lucasfcosta** [comment]
I think @meeber's point makes sense.
I'd like to keep important major versions (`Argon`) explicit.
Since `lts` and `node` are both moving targets in which we always want to have tests on, I'd keep them too.

IMO the latest suggestion by @meeber seems ideal: both explicit and complete.

**Comment by meeber** [comment]
Formalizing my thoughts, mostly for future reference:

- If there was only ever a total of two versions actively supported, one LTS and one Current, at all times, then the `lts/*` and `node` aliases would alone be sufficient. The challenge comes from the planned overlaps that are built into the Node release schedule, causing there to periodically be two LTS or two Current versions actively supported at once. The only way to prevent an actively supported version from being prematurely dropped from automated testing during these periods of overlap is to manually list all actively supported versions. The aliases `lts/*` and `node` should still be included as safety nets, ensuring that there's still good (but not perfect) coverage if package maintainers fail to keep on top of manually updating the version list. (Need to verify that if the same version is listed twice, once numerically and once with an alias, it's only tested by Travis once.)

- Whether or not to use numeric versions like `4` or aliased versions like `lts/argon` is merely preference. I prefer the former because I think it more accurately reflects how most people discuss and interact with different Node versions.

- Including expiration date comments for each version in `.travis.yml` is good.

**Comment by meeber** [comment]
@lucasfcosta @keithamus I just found out that including a release's numeric version and aliased version will cause Travis to run the test suite twice for the same release. Not sure if this behavior is actually problematic in any meaningful way, but it's worth noting.

**Comment by keithamus** [comment]
We've settled on this now, in other repos haven't we @meeber?

@amilajack do you feel happy to update this PR to something like [this](https://github.com/chaijs/deep-eql/blob/d5aedb69c9924e0f6c49ff87d8e6dfb2f5232fe4/.travis.yml#L8-L14)?

**Comment by keithamus** [comment]
FYI @amilajack - I'm closing this one because I've rebased it to reflect our desired changes - in a new PR over at #96. Thanks for all your efforts here. Sorry it took so long to get this merged!

**Pull request closed without merging**